src/vision/status_quo/barbara_bridges_sync_and_async.md

```diff
@@ -0,0 +1,245 @@
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!
+
+## The story
+
+### Introducing `block_on`
+
+Barbara is working on the code for [perf.rust-lang.org](https://perf.rust-lang.org/) and she wants to do a web request to load various intermediate results. She has heard that the `reqwest` crate is quite nice, so she decides to give it a try. She writes up an async function that does her web request:
+
+```rust
+async fn do_web_request(url: &Url) -> Data {
+    ...
+}
+```
+
+She needs to apply this async function to a number of urls. She wants to use the iterator map function, like so:
+
+```rust
+async fn do_web_request(url: &Url) -> Data {...}
+
+fn aggregate(urls: &[Url]) -> Vec<Data> {
+    urls
+        .iter()
+        .map(|url| do_web_request(url))
+        .collect()
+}
+
+fn main() {
+    /* do stuff */
+    let data = aggregate();
+    /* do more stuff */
+}
+```
+
+Of course, since `do_web_request` is an async fn, she gets a type error from the compiler:
+
+```
+error[E0277]: a value of type `Vec<Data>` cannot be built from an iterator over elements of type `impl Future`
+  --> src/main.rs:11:14
+   |
+11 |             .collect();
+   |              ^^^^^^^ value of type `Vec<Data>` cannot be built from `std::iter::Iterator<Item=impl Future>`
+   |
+   = help: the trait `FromIterator<impl Future>` is not implemented for `Vec<Data>`
+```
+
+"Of course," she thinks, "I can't call an async function from a closure." She decides that since she is not overly concerned about performance, so she decides she'll just use a call to [`block_on` from the `futures` crate](https://docs.rs/futures/0.3.14/futures/executor/fn.block_on.html) and execute the function synchronously:
+
+```rust
+async fn do_web_request(url: &Url) -> Data {...}
+
+fn aggregate(urls: &[Url]) -> Vec<Data> {
+    urls
+        .iter()
+        .map(|url| futures::executor::block_on(do_web_request(url)))
+        .collect()
+}
+
+fn main() {
+    /* do stuff */
+    let data = aggregate();
+    /* do more stuff */
+}
+```
+
+The code compiles, and it seems to work.
+
+### Introducing `block_on`
+
+As Barbara works on play, she realizes that she needs to do more and more async operations. She decides to convert her synchronous `main` function into an `async main`. She's using tokio, so she is able to do this very conveniently with the `#[tokio::main]` decorator:
+
+```rust
+#[tokio::main]
+async fn main() {
+    /* do stuff */
+    let data = aggregate();
+    /* do more stuff */
+}
+```
+
+Everything seems to work ok on her laptop, but when she pushes the code to production, it deadlocks immediately. "What's this?" she says. Confused, she runs the code on her laptop a few more times, but it seems to work fine.
```

Comment:
I should maybe add a FAQ to explain this. What's going on here is that her laptop has multiple cores, so the problem with `block_on` doesn't show up. The production server has only a single core because it's a lightweight virtual machine.

Reply:
> I should maybe add a FAQ to explain this. What's going on here is that her laptop has multiple cores, so the problem with block_on doesn't show up. The production server has only a single core because it's a lightweight virtual machine.

I experienced exactly the same thing. It can be verified on the local machine IIRC with `#[tokio::main(flavor = "multi_thread", worker_threads = 1)]`. I also asked in the channel why this isn't enabled by default in dev builds. If we enable something like this, it could mask other concurrency bugs that involves multiple workers.

Reply:
I added a FAQ. Regarding your experience of figuring out where it was stuck, I would say that this is [Barbara wants async insights](https://rust-lang.github.io/wg-async-foundations/vision/status_quo/barbara_wants_async_insights.html)